drivers/usb/udc/udc_renesas_ra.c

```diff
@@ -0,0 +1,756 @@
+		break;
+	}
+}
+
+static void usb_interrupt_handler(void *arg)
+{
+	ARG_UNUSED(arg);
+	usb_device_isr();
+}
+
+static void udc_event_xfer_next(const struct device *dev, const uint8_t ep)
+{
+	struct udc_renesas_ra_data *data = udc_get_private(dev);
+	struct net_buf *buf;
+
+	if (udc_ep_is_busy(dev, ep)) {
+		return;
+	}
+
+	buf = udc_buf_peek(dev, ep);
+	if (buf != NULL) {
+		int err;
+
+		if (USB_EP_DIR_IS_IN(ep)) {
+			err = R_USBD_XferStart(&data->udc, ep, buf->data, buf->len);
+		} else {
+			err = R_USBD_XferStart(&data->udc, ep, buf->data, buf->size);
+		}
+
+		if (err != FSP_SUCCESS) {
+			LOG_ERR("ep 0x%02x error", ep);
+			/* REVISE: remove from endpoint queue? ASSERT? */
+			udc_submit_ep_event(dev, buf, -ECONNREFUSED);
+		} else {
+			udc_ep_set_busy(dev, ep, true);
+		}
+	}
+}
+
+static int usbd_ctrl_feed_dout(const struct device *dev, const size_t length)
+{
+	struct udc_ep_config *cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
+	struct net_buf *buf;
+	struct udc_renesas_ra_data *data = udc_get_private(dev);
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, length);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	k_fifo_put(&cfg->fifo, buf);
+
+	return FSP_SUCCESS == R_USBD_XferStart(&data->udc, cfg->addr, buf->data, buf->size) ? 0
+											    : -EIO;
+}
+
+static int udc_event_xfer_setup(const struct device *dev, struct udc_renesas_ra_evt *evt)
+{
+	struct net_buf *buf;
+	int err;
+
+	struct usb_setup_packet *setup_packet =
+		(struct usb_setup_packet *)&evt->hal_evt.setup_received;
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, sizeof(struct usb_setup_packet));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate for setup");
+		return -ENOMEM;
+	}
+
+	udc_ep_buf_set_setup(buf);
+	memcpy(buf->data, setup_packet, sizeof(struct usb_setup_packet));
+	net_buf_add(buf, sizeof(struct usb_setup_packet));
+
+	/* Update to next stage of control transfer */
+	udc_ctrl_update_stage(dev, buf);
+
+	if (udc_ctrl_stage_is_data_out(dev)) {
+		/*  Allocate and feed buffer for data OUT stage */
+		LOG_DBG("s:%p|feed for -out-", buf);
+		err = usbd_ctrl_feed_dout(dev, udc_data_stage_length(buf));
+		if (err == -ENOMEM) {
+			err = udc_submit_ep_event(dev, buf, err);
+		}
+	} else if (udc_ctrl_stage_is_data_in(dev)) {
+		err = udc_ctrl_submit_s_in_status(dev);
+	} else {
+		err = udc_ctrl_submit_s_status(dev);
+	}
+
+	return err;
+}
+
+static void udc_event_xfer_ctrl_in(const struct device *dev, struct net_buf *const buf)
+{
+	struct udc_renesas_ra_data *data = udc_get_private(dev);
+
+	if (udc_ctrl_stage_is_status_in(dev) || udc_ctrl_stage_is_no_data(dev)) {
+		/* Status stage finished, notify upper layer */
+		udc_ctrl_submit_status(dev, buf);
+	}
+
+	udc_ctrl_update_stage(dev, buf);
+
+	if (udc_ctrl_stage_is_status_out(dev)) {
+		/* Status stage finished, notify upper layer */
+		udc_ctrl_submit_status(dev, buf);
+	}
+
+	/* Clear setup endpoint. Send a ZLP */
+	R_USBD_XferStart(&data->udc, USB_CONTROL_EP_IN, NULL, 0);
```

Comment:
Are you sure that sending ZLP for every control transfer is needed?

Reply:
I fixed it